File: Array/zhanchengguo/0616-有效的括号/Main.java
package com.hlg.daydaytobusiness;

import android.util.Log;

import java.util.Stack;

public class Main {

    void main() {
        boolean result = isValid("()[]{}[");
        Log.e("test", "value is " + result);
    }

    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if ('(' == c || '[' == c || '{' == c) {
                stack.push(c);
            } else {
                if (stack.isEmpty()) {
                    return false;
                }
                char top = stack.peek();
                stack.pop();
                if (('(' == top && c != ')') ||
                        ('[' == top && c != ']') ||
                        ('{' == top && c != '}')) {
                    return false;
                }
            }
        }
        return stack.isEmpty();
    }
}

File: Array/zhanchengguo/0616-有效的括号/main.kt
package com.hlg.daydaytobusiness

import java.util.*

fun main() {
    val result = isValid("()[]{}[")
}

fun isValid(s: String): Boolean{
    val stack = Stack<Char>()
    for (i in 0 until s.length) {
        val c = s[i]
        if ('(' == c || '[' == c || '{' == c) {
            stack.push(c)
        } else {
            if (stack.isEmpty()) {
                return false
            }
            val top = stack.peek()
            stack.pop()
            if ('(' == top && c != ')' ||
                    '[' == top && c != ']' ||
                    '{' == top && c != '}') {
                return false
            }
        }
    }
    return stack.isEmpty()
}